explorer/interpreter/resolve_names.cpp

```diff
@@ -11,22 +11,137 @@
+                       bool allow_qualified_names = false) -> ErrorOr<Success>;
+
+  // Traverse the sub-AST rooted at the given node, resolve all names within
+  // it using enclosing_scope, and update enclosing_scope to add names to
+  // it as they become available. In scopes where names are only visible below
+  // their point of declaration (such as block scopes in C++), this is
+  // implemented as a single pass, recursively calling ResolveNames on the
+  // elements of the scope in order. In scopes where names are also visible
+  // above their point of declaration (such as class scopes in C++), this
+  // requires three passes: first calling AddExposedNames on each element of the
+  // scope to populate a StaticScope, and then calling ResolveNames on each
+  // element, passing it the already-populated StaticScope but skipping member
+  // function bodies, and finally calling ResolvedNames again on each element,
+  // and this time resolving member function bodies.
+  auto ResolveNames(Expression& expression, const StaticScope& enclosing_scope)
+      -> ErrorOr<Success>;
+  auto ResolveNames(WhereClause& clause, const StaticScope& enclosing_scope)
+      -> ErrorOr<Success>;
+  auto ResolveNames(Pattern& pattern, StaticScope& enclosing_scope)
+      -> ErrorOr<Success>;
+  auto ResolveNames(Statement& statement, StaticScope& enclosing_scope)
+      -> ErrorOr<Success>;
+  auto ResolveNames(Declaration& declaration, StaticScope& enclosing_scope,
+                    ResolveFunctionBodies bodies) -> ErrorOr<Success>;
+
+  auto ResolveMemberNames(llvm::ArrayRef<Nonnull<Declaration*>> members,
+                          StaticScope& scope, ResolveFunctionBodies bodies)
+      -> ErrorOr<Success>;
+
+ private:
+  // Mapping from namespaces to their scopes.
+  llvm::DenseMap<const NamespaceDeclaration*, StaticScope> namespace_scopes_;
+
+  // Mapping from declarations to the scope in which they expose a name.
+  llvm::DenseMap<const Declaration*, StaticScope*> exposed_name_scopes_;
+};
+
+}  // namespace
+
+auto NameResolver::ResolveQualifier(DeclaredName name,
+                                    StaticScope& enclosing_scope,
+                                    bool allow_undeclared)
+    -> ErrorOr<Nonnull<StaticScope*>> {
+  Nonnull<StaticScope*> scope = &enclosing_scope;
+  for (auto& [loc, qualifier] : name.qualifiers()) {
+    // TODO: If we permit qualified names anywhere other than the top level, we
+    // will need to decide whether the first name in the qualifier is looked up
+    // only in the innermost enclosing scope or in all enclosing scopes.
+    CARBON_ASSIGN_OR_RETURN(
+        ValueNodeView node,
+        scope->ResolveHere(qualifier, loc, allow_undeclared));
+
+    if (auto* namespace_decl = dyn_cast<NamespaceDeclaration>(&node.base())) {
+      scope = &namespace_scopes_[namespace_decl];
+    } else {
+      return ProgramError(name.source_loc())
+             << "qualifier `" << qualifier << "` is not a namespace";
```

Comment:
Isn't "is not a namespace" too specific? In the test: `fn F.G() {}`, `F` could also be a class, interface, etc and that could still be valid. Maybe instead say something more like: `F is a function, and cannot be a qualifier for declarations.`

Reply:
It's correct for now, compared to explorer's current behavior. The immediately-surrounding code is explicitly dealing with namespaces, so I'm not concerned that we might forget to update the diagnostic when that changes.

That said, making the diagnostic say what the name resolved to does seem nice. Ideally I'd want to point to its location in the source code, but our diagnostics infrastructure in explorer isn't well-suited to that. How would you feel about including the `PrintID` output here? Eg, `fn F cannot be used as a name qualifier` or similar? I'm not sure that it makes sense to invent a new mapping from declaration kinds to end-user-visible strings for just this one diagnostic.

Reply:
PrintID seems fine. I think there'd been a PR trying to add stringification of Value.Kind for some other purpose, but I don't view it as that important. (if you changed Value.Kind to use EnumBase, you might get less-than-ideal stringification for free too)